### PR TITLE
Fix for DROOLS-975 - added TransactionManagerFactory

### PR DESCRIPTION
**Pull request by lukenjmcd**
Added TransactionManagerFactory to remove coupling between Drools/jBPM classes and the JtaTransactionManager.

### CODE REVIEW

File: drools-persistence-jpa/src/main/java/org/drools/persistence/TransactionManagerFactory.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * 
+ * The implementation class can be specified using the {@code opt.kie.tmfactory.class} system property.
+ * 
+ */
+public abstract class TransactionManagerFactory {
+    private static final Logger logger = LoggerFactory.getLogger(TransactionManagerFactory.class);
+    private static volatile TransactionManagerFactory INSTANCE;
+
+    static {
+        String factoryClassName = System.getProperty("opt.kie.tmfactory.class", JtaTransactionManagerFactory.class.getName());
```
Comment (mswiderski) [inline drools-persistence-jpa/src/main/java/org/drools/persistence/TransactionManagerFactory.java]: could we change the name of system property to be consistent with other system properties to something like this

```
org.kie.txm.factory.class
```

**Comment by mswiderski** [comment]
ok to test

**Comment by mswiderski** [comment]
in general looks good to me, one question - have you looked at droolsjbpm-integration repository to check if there is no need for any other alignments after introducing this factory? Important parts there are:
- kie-spring
- drools-osgi
  so might be worth double checking if it does work there the way it was designed - with factory

File: drools-persistence-jpa/src/main/java/org/drools/persistence/TransactionManagerFactory.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            logger.error("Unable to instantiate "+factoryClassName, e);
+        }
+    }
+    
+    /**
+     * Sets the factory used for creating {@link TransactionManager}s.
+     * 
+     * @param factory
+     */
+    public static final void set(TransactionManagerFactory factory) {
```
Comment (mrietveld) [inline drools-persistence-jpa/src/main/java/org/drools/persistence/TransactionManagerFactory.java]: Please remove the `public static final set` method: IMHO, this is far too dangerous and unneccessary, AFAICT.

**Comment by lukenjmcd** [comment]
A couple questions (this is my first time using github or submitting to an open-source project) 

1 - is it my responsibility to make the changes being proposed here (change property name, remove set() method)?
2 - if yes to question 1, do I need to create a new pull request with those changes or is there some documentation about the process to follow for this?

Thanks.

**Comment by psiroky** [comment]
@lukenjmcd it would be great if you could do those suggested changes. There is no need to create new PR, you can amend (update) this one. See for example http://blog.adamspiers.org/2015/03/24/why-and-how-to-correctly-amend-github-pull-requests/ for more details on how to do that.

**Comment by mrietveld** [comment]
@lukenjmcd It's enough to modify your local branch and then `git push -f origin <pr-branch-name>`, where origin is your github repository name. (I'm assuming that "upstream" is the alias for droolsjbpm github repository).

**Comment by lukenjmcd** [comment]
I've made the suggested changes.  I received a failure notification but I get "server not found" error when trying to see the details.  The tests pass on my local system.  

I also made a PR for a change in droolsjbpm-integration/kie-infinispan/drools-infinispan-persistence

**Comment by mariofusco** [comment]
Merged by https://github.com/droolsjbpm/drools/commit/d67f775502821d40be9404e7d3ced44f9976e97c

**Pull request closed without merging**